### PR TITLE
[skia] Fix include path

### LINKED ISSUE #32657: [skia] Include Path should point to include/ instead of include/skia
### Is your feature request related to a problem? Please describe.

The described problem was encountered on a Windows dev machine with Visual Studio 22 as build environment and CMake 3.21.4.  

We include skia vcpkg port in own projects using CMake. When including it the recommended way ...

```
...
find_package(unofficial-skia CONFIG REQUIRED)
...
target_link_libraries(myProject PRIVATE unofficial::skia::skia)
```
then CMake generates visual studio projects with an additional include files directory pointing to  
`...\vcpkg\installed\x64-windows\include\skia`.

So in cpp source code using skia we have to include skia header files like this:

```
#include <core/SkSurface.h>
```

A better way would be to be able to include skia header files like this 
```
#include <skia/core/SkSurface.h>
```
This is also the way other vcpkg ports like gtest do it. This would require the aforementioned additional include directory to point to `...\vcpkg\installed\x64-windows\include\`. Please consider updating the port in this way.

**Additonal Note**
When using the "skia"-package instead of "unofficial-skia", e.g. 
```
find_package(skia CONFIG REQUIRED)
...
target_link_libraries(myProject PRIVATE skia::skia)
```
then the include directory points to  `...\vcpkg\installed\x64-windows\include\` as desired. But CMake generates warnings with that configuration:
```
  find_package(skia) is deprecated.
  skia provides CMake targets:
      find_package(unofficial-skia CONFIG REQUIRED)
      target_link_libraries(main PRIVATE unofficial::skia::skia)
```

### Proposed solution

When integration skia vcpkg port in CMake via

```
...
find_package(unofficial-skia CONFIG REQUIRED)
...
target_link_libraries(myProject PRIVATE unofficial::skia::skia)
```

please make the include directory in CMake-generated C++ projects point to `...\vcpkg\installed\x64-windows\include\` instead of `...\vcpkg\installed\x64-windows\include\skia\`.

### Describe alternatives you've considered

Including skia header files like this
```
#include <core/SkSurface.h>
```

which makes it hard to understand to which library the included file belongs to. 

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by xiaozhuai**
- [x] Changes comply with the [maintainer guide](https://github.com/microsoft/vcpkg-docs/blob/main/vcpkg/contributing/maintainer-guide.md)
- [x] SHA512s are updated for each updated download
- [x] The "supports" clause reflects platforms that may be fixed by this new version
- [x] Any fixed [CI baseline](https://github.com/microsoft/vcpkg/blob/master/scripts/ci.baseline.txt) entries are removed from that file.
- [x] Any patches that are no longer applied are deleted from the port's directory.
- [x] The version database is fixed by rerunning `./vcpkg x-add-version --all` and committing the result.
- [x] Only one version is added to each modified port's versions file.

Fix #32657 and #28591
It was a mistake in the first place, and now it's time to fix it.

### CODE REVIEW

**Comment by dg0yt** [comment]
> It was a mistake in the first place, and now it's time to fix it.

Opinion, not fact.

**Comment by BillyONeal** [comment]
I don't have a horse in this race, but I note there appears to be only one downstream consumer in the repo, `msdfgen`, and they say `#include <skia/core/SkPath.h>`.

**Comment by xiaozhuai** [comment]
> >It was a mistake in the first place, and now it's time to fix it.

>Opinion, not fact.

There have been debates on this issue before, so I don't want to repeat myself.
I don't care if this PR ends up being merged.
But I'm sure there will be more reports about the same problem.

**Review by Adela0814 — DISMISSED** [review]

**Comment by BillyONeal** [comment]
@xiaozhuai can you clarify the problem that this is actually fixing? I did the 'data gathering' above but as I said I don't have strong opinions here; I'm concerned that this is going to break existing users for no obvious benefit. (Both `#include <skia/blah.h>` and `#include <blah.h>` work with the status quo)

**Comment by xiaozhuai** [comment]
@BillyONeal 
> I'm concerned that this is going to break existing users for no obvious benefit.
It won't. As you said, Both `#include <skia/blah.h>` and `#include <blah.h>` work.

**Comment by BillyONeal** [comment]
> As you said, Both `#include <skia/blah.h>` and `#include <blah.h>` work after this change.

Hmmm that was not my understanding of the change; my understanding was that today they both work, and after the change only `<skia/blah.h>` works. If they both worked before and both work now what is actually changed here?

>Finally, take into account not breaking existing users. This PR does not remove the second way.
At the same time, those who prefer the first way are given the right to choose.

I'm confused, how does the user get to 'choose' as a result of this change?

**Comment by xiaozhuai** [comment]
> Hmmm that was not my understanding of the change; my understanding was that today they both work, and after the change only <skia/blah.h> works.

No, currentlly only `blah.h` works. This pr make them both work. Please review this pr again.

**Comment by dg0yt** [comment]
@BillyONeal The port installs two configs: The recommended config with the `unofficial` prefix only supports `include/skia`. The legacy config also supported `include`, to not break legacy users. This PR move this legacy property to the recomended config.

What I dislike about this change is that it basically makes all headers in `include` available to consumers which link to skia. 
This is not breaking users or ports *now* but it may hide dependencies, causing unexpected problems *later*. 
The problem is not so much with port skia (which resides in `include/skia`) but with all those other ports which install headers directly to `include`. For example, we recently fixed `libkml` issues around `include/zip.h` being used unexpectedly instead of `minizip/zip.h`.

Skias's public headers are sufficiently distinct by using an `Sk` filename component. This is somewhat similar to the `Q` in Qt's headers. That's why I think we should avoid adding the general include path unless it becomes the *official* pattern to use skia headers. Users which want to refer to headers with a `skia` prefix may add this special behavior in their projects by inspecing the properties of the unofficial cmake config.

That's why I say the current implementation of the unofficial config is not by mistake, or a bug. But I don't claim I know the official pattern, or even the most popular one. And even if the official pattern would be `#include <skia/core/SkSurface.h>`, the best fix might be to install the headers to `include/skia/skia` instead - if there weren't user which don't use cmake config at all...

**Comment by dg0yt** [comment]
> What problem?

I named a real example.
I had another input to move forward to a good decision, but I prefer to leave this discussion at this point.

**Comment by xiaozhuai** [comment]
There is no point in arguing any more, and I will not reply to anything more on this.

Let the team decide.

**Comment by BillyONeal** [comment]
@dg0yt :

> The port installs two configs: The recommended config with the `unofficial` prefix only supports `include/skia`. The legacy config also supported `include`, to not break legacy users. This PR move this legacy property to the recomended config.
> [...]
> That's why I say the current implementation of the unofficial config is not by mistake, or a bug. But I don't claim I know the official pattern, or even the most popular one. And even if the official pattern would be `#include <skia/core/SkSurface.h>`, the best fix might be to install the headers to `include/skia/skia` instead - if there weren't user which don't use cmake config at all...

I see, thank you for clarifying.

> What I dislike about this change is that it basically makes all headers in `include` available to consumers which link to skia.

I think this is going to be the case regardless. Most ports expect 'one include directory to rule them all', and this is in fact the way a lot of downstream customers can get to dependencies at all. So I think this ship has sailed.

> The problem is not so much with port skia (which resides in `include/skia`) but with all those other ports which install headers directly to `include`. For example, we recently fixed `libkml` issues around `include/zip.h` being used unexpectedly instead of `include/minizip/zip.h`.

But if they said `#include <minizip/zip.h>`, wouldn't that be a better resolution of this problem? With the proposed behavior `libkml` is still toast if more than one thing tries to provide `zip.h`. On the contrary, the behavior here can hide the problem while vcpkg's normal 'only one owner' file checks would catch this.

@xiaozhuai :

> There are many ports behave like this. If this becomes a problem, then the vcpkg installation directory should be redesigned.

This is not a vcpkg problem, this is a C++ problem. No 2 entities can own the same `#include <stuff>` name; the C++ compiler is only going to pick one of them.

**Comment by dg0yt** [comment]
> Most ports expect 'one include directory to rule them all', and this is in fact the way a lot of downstream customers can get to dependencies at all. So I think this ship has sailed.

At least this ship hasn't sailed yet for the `unofficial-skia` config package.
And in this particular context, it should be noted that skia/gn/bazel try hard to do the opposite of 'one include directory to rule them all', to avoid problems vcpkg has to fight with.

> But if they said `#include <minizip/zip.h>`, wouldn't that be a better resolution of this problem?

This is the ship which has sailed: Official usage is `#include <zip.h>`. The generla pattern of this problem is the lack of disambiguation by unique components. This would be mitigated by avoiding the 'one include directory to rule them all' approach.
But skia headers have a unique component: the `Sk` name prefix.

> No 2 entities can own the same `#include <stuff>` name; the C++ compiler is only going to pick one of them.

I would say that a single translation unit cannot use two different instances of `<stuff>`. But it is not to hard to use different `<stuff>` for different CMake targets, in a controlled way.

---

Last not least: The `skia` path component is a proprietary vcpkg port property.
- Upstream doesn't support installation at all. gn or bazel projects use subdirectories for dependencies (as does skia, making it more difficult to devendor dependencies).
- Upstream keeps public headers in `include`. 
  (Unfortunately, some headers are needed also from `modules`.)
 
So expected usage patters are more like
~~~CXX
#include "include/codec/SkCodec.h"
~~~
as in `convert-to-nia.cpp` which is used for the test port. 
And that's why the port's primary installation dir is `include/skia/include`. This content is duplicated into `include/skia` to support the legacy config as long as we want to maintain support for it. Once we drop the legacy config, I would drop the duplication. (Consumers which didn't use the `include/` path component can be supported by adding `include/skia/include` as another item in the interface include directories.)

FTR [Chrome uses](https://github.com/chromium/chromium/blob/985e20f7e023b781380cee10f803604eb425f471/ui/base/clipboard/scoped_clipboard_writer.h#L13)
~~~CXX
#include "third_party/skia/include/core/SkBitmap.h"
~~~
where `skia` is the source code directory (parent dir of `include`), not a subdir of `include`. So if users want to work with a `skia` path component (after this PR!), they should actually spell it out like
~~~CXX
#include <skia/include/codec/SkBitmap.h>
~~~

---

To sum it up, I don't mean to block a change. But I would like to make it clear that
- the current config is not a bug or mistake but was chosen for good reasons,
- the existence of a `skia` path component is more or less specific to vcpkg.
- the proposed usage of `#include <skia/core/SkBitmap.h>` will make it impossible to eventually drop the duplicated installation.

**Comment by xiaozhuai** [comment]
> the proposed usage of #include <skia/core/SkBitmap.h> will make it impossible to eventually drop the duplicated installation.

It was fixed.
https://github.com/microsoft/vcpkg/pull/32660/commits/bf1341eed1a03f1c4b9c4140e91911c7af5e45ec.

**Comment by BillyONeal** [comment]
@AugP @markle11m @dan-shaw @JavierMatosD @ras0219-msft and I discussed this today.

* The vcpkg team considers 'hiding' headers with additional `-I`s, as seemed to be @dg0yt's design
  intent here, a non-goal.
* @ras0219-msft did a bunch of github searches which seem to indicate the most common use is
  1. forks of skia itself
  2. `#include "include/core/skcanvas.h"`
  which suggests that, as @dg0yt says, the subdirectory `<skia` is a vcpkg fabrication and that we
  should actually be installing skia's headers to `${CURRENT_PACKAGES_DIR}/include/include` (as in,
  make `"${CURRENT_PACKAGES_DIR}/include/include/core/SkCanvas.h"`)
* Yes, this means we should fix `msdfgen` to say `#include <include/core/SkPath.h>` rather than
  `#include <skia/core/SkPath.h>`
* It would be nice if we have some way to contact upstream and see what they think. Unfortunately it seems they only care about Bazel :(

**Comment by xiaozhuai** [comment]
@BillyONeal 

> Yes, this means we should fix msdfgen to say #include <include/core/SkPath.h> rather than
#include <skia/core/SkPath.h>

Since this PR can't be merged, let's  leave the status quo, at least without breaking downstream users.
Before there is an official statement, `#include <include/core/SkPath.h>` is also a speculation.
IMO, change to `#include <include/core/SkPath.h>` won't help much. Instead, it can lead to misunderstandings.

So be it, I'm going to close this PR.
